### ISSUE TITLE
"Gradient accumulation on training samples of variable size" example broken when using DeepSpeed

### ISSUE THREAD

**Issue by inkitori** [huggingface/accelerate#3780]
I've been stuck for a while trying to get gradient accumulation and multi-GPU training with DeepSpeed to work. I noticed that even when I keep my effective batch sizes the same, if I do one run with 2 GPUs and the other with 4 the losses are completely different. 

I thought that this could be just a bug in my own code so I tried running only the [example](https://huggingface.co/docs/accelerate/en/usage_guides/gradient_accumulation#self-contained-causal-lm-example) provided on gradient accumulation for causal LM training, but even when I ran it with and without DeepSpeed I got the following outputs:

```
Device 1 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
Device 3 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
Device 0 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
Device 2 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
w/o accumulation, the final model weight is tensor([0.4339, 0.0951])
```

Is the gradient accumulation documentation outdated? And if so, what do I need to change in my code to fix it? I'm not sure if it's also an issue with DeepSpeed either not using the gradient_accumulation_steps I pass into Accelerator, DeepSpeed not properly dividing the loss by gradient_accumulation_steps when I call accelerator.backward(), or the losses being divided by num_processes. Thanks!

**Comment by inkitori** [comment]
Sorry, it turns out it does work without DeepSpeed at least initially. However, if DeepSpeed has been enabled the example breaks with or without DeepSpeed. Below is my config for reference:


```
compute_environment: LOCAL_MACHINE
debug: false
deepspeed_config:
  zero3_init_flag: false
  zero_stage: 2
distributed_type: DEEPSPEED
downcast_bf16: 'no'
enable_cpu_affinity: false
machine_rank: 0
main_training_function: main
mixed_precision: fp16
num_machines: 1
num_processes: 4
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

And here are the example logs. These were with DeepSpeed enabled, but even after disabling DeepSpeed, it outputs the exact same logs.

```
initial model weight is tensor([-0.0075,  0.5364])
initial model clone weight is tensor([-0.0075,  0.5364])
Step 0 - Device 2 - num items in the local batch 148
Step 0 - Device 0 - num items in the local batch 84
Step 0 - Device 1 - num items in the local batch 116
Step 0 - Device 3 - num items in the local batch 180
Total num items 528
[2025-09-14 21:55:55,437] [INFO] [real_accelerator.py:260:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2025-09-14 21:55:55,437] [INFO] [real_accelerator.py:260:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2025-09-14 21:55:55,437] [INFO] [real_accelerator.py:260:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2025-09-14 21:55:55,437] [INFO] [real_accelerator.py:260:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2025-09-14 21:56:34,117] [INFO] [logging.py:107:log_dist] [Rank -1] [TorchCheckpointEngine] Initialized with serialization = False
[2025-09-14 21:56:34,117] [INFO] [logging.py:107:log_dist] [Rank -1] [TorchCheckpointEngine] Initialized with serialization = False
[2025-09-14 21:56:34,117] [INFO] [logging.py:107:log_dist] [Rank -1] [TorchCheckpointEngine] Initialized with serialization = False
[2025-09-14 21:56:34,117] [INFO] [logging.py:107:log_dist] [Rank -1] [TorchCheckpointEngine] Initialized with serialization = False
Device 1 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
Device 0 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
Device 3 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
Device 2 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
```

Here's another example where I increased the dataset size:

```
initial model weight is tensor([-0.0075,  0.5364])
initial model clone weight is tensor([-0.0075,  0.5364])
Step 0 - Device 1 - num items in the local batch 12680
Step 0 - Device 3 - num items in the local batch 13320
Step 0 - Device 2 - num items in the local batch 13000
Step 0 - Device 0 - num items in the local batch 12360
Total num items 51360
[2025-09-14 22:02:20,211] [INFO] [real_accelerator.py:260:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2025-09-14 22:02:20,211] [INFO] [real_accelerator.py:260:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2025-09-14 22:02:20,211] [INFO] [real_accelerator.py:260:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2025-09-14 22:02:20,211] [INFO] [real_accelerator.py:260:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2025-09-14 22:02:59,354] [INFO] [logging.py:107:log_dist] [Rank -1] [TorchCheckpointEngine] Initialized with serialization = False
[2025-09-14 22:02:59,354] [INFO] [logging.py:107:log_dist] [Rank -1] [TorchCheckpointEngine] Initialized with serialization = False
[2025-09-14 22:02:59,354] [INFO] [logging.py:107:log_dist] [Rank -1] [TorchCheckpointEngine] Initialized with serialization = False
[2025-09-14 22:02:59,354] [INFO] [logging.py:107:log_dist] [Rank -1] [TorchCheckpointEngine] Initialized with serialization = False
Device 2 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
Device 1 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
Device 0 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
Device 3 - w/ accumulation, the final model weight is tensor([-0.0075,  0.5364])
w/o accumulation, the final model weight is tensor([ 4.2845, -3.7555])
```
